scalding-core/src/main/scala/com/twitter/scalding/Job.scala

```diff
@@ -498,6 +498,29 @@ class ScriptJob(cmds: Iterable[String]) extends Job(Args("")) {
   }
 }
 
+/**
+ * Allows custom counter verification logic when the job completes.
+ */
+trait CounterVerification extends Job {
+  /**
+   * Verify counter values. The job will fail if this returns false or throws an exception.
+   */
+  def verifyCounters(counters: Map[StatKey, Long]): Boolean
+
+  /**
+   * Override this to false to skip counter verification in tests.
+   */
+  def verifyCountersInTest: Boolean = true
+
+  override def listeners: List[FlowListener] = {
+    if (this.verifyCountersInTest) {
+      super.listeners :+ new StatsFlowListener(this.verifyCounters)
+    } else {
+      super.listeners
+    }
```

Comment:
This needs to check for `TestMode` along with the flag here, right? and always add `StatsFlowListener` in other modes.